Skip function-like and anonymous class nodes when checking if a function or method is a generator

sebkehr opened the pull request:
Fixes: phpstan/phpstan#12462.

As is, when checking if a function (or method) is a generator (via `PhpFunctionFromParserNodeReflection::isGenerator()`) all subnodes are recursively checked for containment of a yield expression. Hence any function or method itself returning a yielding closure or arrow function is falsely determined to be a generator which in turn leads to incorrect return type checks as demonstrated in https://phpstan.org/r/5970e2e2-0aad-4927-8938-3f7d83e73a95.

The proposed fix therefore simply skips `FunctionLike` nodes in the above mentiond recursive check.

Commit:
skip function-like nodes when checking if a function is a generator

ondrejmirtes commented on src/Reflection/Php/PhpFunctionFromParserNodeReflection.php:
```diff
@@ -281,7 +281,10 @@ private function nodeIsOrContainsYield(Node $node): bool
 
-			if ($nodeProperty instanceof Node && $this->nodeIsOrContainsYield($nodeProperty)) {
+			if ($nodeProperty instanceof Node &&
+				!$nodeProperty instanceof FunctionLike &&
```
Would be great to also not descend into anonymous classes.

sebkehr replied:
@ondrejmirtes Anonymous class nodes should now be skipped as well. I submitted the fix via Roave/BetterReflection#1479.

ondrejmirtes reviewed (changes requested):
Please also submit the same fix to https://github.com/Roave/BetterReflection: https://github.com/Roave/BetterReflection/blob/d262672652b9b70a54246adadf340587b535f549/src/Reflection/ReflectionFunctionAbstract.php#L349-L380

ondrejmirtes reviewed (changes requested):
I just realized there's one more place that needs fixing. Check out CleaningParserTest. The job of CleaningParser is to erase function bodies of not-analyzed code (like in vendor/) before the AST is cached into memory.

I worry the code is faulty and it might do something like this:

Before:

```php
function foo() {
    $fn = fn () => yield;
};
```

After:

```php
function foo() {
    yield;
};
```

Changing function foo() from non-generator to a generator.

Thank you!

sebkehr commented:
> I just realized there's one more place that needs fixing. Check out CleaningParserTest. The job of CleaningParser is to erase function bodies of not-analyzed code (like in vendor/) before the AST is cached into memory.

Ah ok, I'll look into it.

Commit:
refactored cleaning parser logic

sebkehr commented:
@ondrejmirtes

I added two more examples to  _tests/PHPStan/Parser/data/cleaning-1-before.php_. Please also have a look on the expected outcome after cleaning in _tests/PHPStan/Parser/data/cleaning-1-after.php_.

With the proposed changes to _src/Parser/CleaningVisitor.php_ the closure in _ContainsClosure::doFoo()_ will now be erased not leading to any left-over yield statements just like the yielding array-function in _ContainsClosure::doBar()_. Is this the intended behavior?

I also added _Baz::someGenerator3()_ containing _yield_ as a subexpression which will now lead to a _yield_ expression after cleaning.

sebkehr commented:
@ondrejmirtes 

1. I'm not sure what to do about the [failing check](https://github.com/phpstan/phpstan-src/actions/runs/13310904103/job/37173046112?pr=3794) as it doesn't seem to related to my changes.
2. Should we maybe put the changes to CleaningVisitor into a separate PR?

sebkehr committed:
Merge branch '2.1.x' into improved_return_type_checks_wrt_yielding_closures
